src/Ide/TreeTransform.hs

```diff
@@ -0,0 +1,124 @@
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Ide.TreeTransform
+  ( Graft, graft, transform, useAnnotatedSource
+  ) where
+
+import           BasicTypes (appPrec)
+import           Control.Monad
+import           Data.Functor.Identity
+import qualified Data.Text as T
+import           Debug.Trace
+import           Development.IDE.Core.RuleTypes
+import           Development.IDE.Core.Rules
+import           Development.IDE.Core.Shake
+import           Development.IDE.Types.Location
+import           GHC hiding (parseExpr)
+import           Generics.SYB
+import           Ide.PluginUtils
+import           Language.Haskell.GHC.ExactPrint
+import           Language.Haskell.GHC.ExactPrint.Parsers
+import           Language.Haskell.LSP.Types
+import           Language.Haskell.LSP.Types.Capabilities (ClientCapabilities)
+import           Outputable
+import           Retrie.ExactPrint hiding (parseExpr)
+import           Text.Regex.TDFA.Text()
+
+
+------------------------------------------------------------------------------
+-- | Get the latest version of the annotated parse source.
+useAnnotatedSource
+    :: String
+    -> IdeState
+    -> NormalizedFilePath
+    -> IO (Annotated ParsedSource)
+useAnnotatedSource herald state nfp = do
+  Just pm <- runAction herald state $ use GetParsedModule nfp
```

Comment:
Why are you assuming `Just` here? The type-checked module is retrieved via `useWithStale`, which means it will succeed even if the module doesn't parse. 

Either change `use` to `useWithStale` here, or handle the `Nothing` case. Probably the latter as I suspect you really don't want to use a stale AST.

Reply:
Good catch!